src/xenia/kernel/xam/xam_ui.cc

```diff
@@ -87,30 +87,24 @@ class MessageBoxDialog : public xe::ui::ImGuiDialog {
 };
 
 // http://www.se7ensins.com/forums/threads/working-xshowmessageboxui.844116/?jdfwkey=sb0vm
-SHIM_CALL XamShowMessageBoxUI_shim(PPCContext* ppc_context,
-                                   KernelState* kernel_state) {
-  uint32_t user_index = SHIM_GET_ARG_32(0);
-  uint32_t title_ptr = SHIM_GET_ARG_32(1);
-  uint32_t text_ptr = SHIM_GET_ARG_32(2);
-  uint32_t button_count = SHIM_GET_ARG_32(3);
-  uint32_t button_ptrs = SHIM_GET_ARG_32(4);
-  uint32_t active_button = SHIM_GET_ARG_32(5);
-  uint32_t flags = SHIM_GET_ARG_32(6);
-  uint32_t result_ptr = SHIM_GET_ARG_32(7);
-  uint32_t overlapped_ptr = SHIM_GET_ARG_32(8);
-
+dword_result_t XamShowMessageBoxUI(dword_t user_index, lpstring_t title_ptr,
+                                   lpstring_t text_ptr, dword_t button_count,
+                                   lpvoid_t button_ptrs, dword_t active_button,
+                                   dword_t flags, lpdword_t result_ptr,
+                                   lpvoid_t overlapped_ptr) {
   std::wstring title;
   if (title_ptr) {
-    title = xe::load_and_swap<std::wstring>(SHIM_MEM_ADDR(title_ptr));
+    title = xe::load_and_swap<std::wstring>(title_ptr);
   } else {
     title = L"";  // TODO(gibbed): default title based on flags?
   }
-  auto text = xe::load_and_swap<std::wstring>(SHIM_MEM_ADDR(text_ptr));
+  auto text = xe::load_and_swap<std::wstring>(text_ptr);
   std::vector<std::wstring> buttons;
   std::wstring all_buttons;
   for (uint32_t j = 0; j < button_count; ++j) {
-    uint32_t button_ptr = SHIM_MEM_32(button_ptrs + j * 4);
-    auto button = xe::load_and_swap<std::wstring>(SHIM_MEM_ADDR(button_ptr));
+    uint64_t button_ptr = button_ptrs.as_array<uint32_t>()[j];
+    button_ptr += (uint64_t)kernel_state()->memory()->virtual_membase();
+    auto button = xe::load_and_swap<std::wstring>((void*)button_ptr);
```

Comment:
This would be better as: `kernel_state()->memory()->TranslateVirtual(button_ptr)`.